lib/shopping_cart.rb

```diff
@@ -0,0 +1,64 @@
+class ShoppingCart
+  attr_reader :name, :capacity, :products
+
+  def initialize(name, capacity)
+    @name = name
+    @capacity = capacity.to_i
+    @products = []
+  end
+
+  def add_product(product)
+    @products << product
+  end
+
+  def details
+    {
+      name: @name,
+      capacity: @capacity
+    }
+  end
+
+  def total_number_of_products
+    sum = 0
+    @products.map do |product|
+      sum += product.quantity
+    end
+    sum
+  end
+
+  def is_full?
+    if total_number_of_products > @capacity
+      true
+    else
+      false
+    end
+  end
+
+  def products_by_category(category)
+    @products.find_all do |product|
+      product.category == category
+    end
+  end
+
+  def percent_occupied
+    ((total_number_of_products / capacity.to_f) * 100).round(2)
+  end
+
+  def sorted_products_by_quantity
+    @products.sort_by do |product|
+      product.quantity
+    end.reverse
+  end
+
+  def product_breakdown
+    breakdown = {}
+    @products.map do |product|
+      if breakdown[product.category] == nil
+      breakdown[product.category] = [product]
+      else
+        breakdown[product.category] = (breakdown[product.category] << product)
+      end
+    end
+    breakdown
+  end
```

Comment:
Awesome job on these last two. Nice use of the sort_by enum. For the product_breakdown method, default hashes are always tough syntactically, but you could do something like this to get the structure set up: 
```
Hash.new {|hash_obj, key| hash_obj[key] = []}
```